people-and-planet-ai/land-cover-classification/serving/main.py

```diff
@@ -12,205 +12,73 @@
+@app.route("/predict/<float(signed=True):lon>/<float(signed=True):lat>/<int:year>")
+def predict(lon: float, lat: float, year: int) -> flask.Response:
     """Gets a prediction from the model.
 
     Args:
-        lat: Latitude of the point of interest.
         lon: Longitude of the point of interest.
-        year: Year of interest, the model takes the median of that year.
+        lat: Latitude of the point of interest.
+        year: Year of interest, a median composite is used.
 
-    Optional request parameters:
-        patch-size: Size in pixels of the surrounding square patch to use.
-        model-path: Cloud Storage path to load the model to use.
+    Optional query parameters:
+        patch-size: Size in pixels of the surrounding square patch.
 
     Returns:
-        The bytes of a compressed numpy file with the results if successful,
-        or the error with a 500 status code otherwise.
+        A JSON response with the predictions if successful, or an error otherwise.
     """
 
     # Optional HTTP request parameters.
     #   https://en.wikipedia.org/wiki/Query_string
-    patch_size = flask.request.args.get("patch-size", 256, type=int)
-    model_path = flask.request.args.get("model-path", DEFAULT_MODEL_PATH)
+    patch_size = flask.request.args.get("patch-size", 512, type=int)
 
-    # If anything fails, send a more descriptive error message instead of
-    # a generic "500: Internal Server Error" message.
     try:
-        ee_init()
-
-        # Fetch the input patch.
-        patch = get_patch(
-            image=sentinel2_image(f"{year}-01-01", f"{year}-12-31"),
-            lat=lat,
-            lon=lon,
-            bands=INPUT_BANDS,
-            patch_size=patch_size,
-            scale=10,
-        )
-
-        # Load the model and get the predictions. If the model is hosted
-        # somewhere else, this is where we would send a request.
-        model = tf.keras.models.load_model(model_path)
-
-        # Create an input dictionary with a batch containing a single patch.
-        inputs_batch = {name: np.stack([patch[name]]) for name in patch.dtype.names}
-
-        # Get the first (and only) element in the predictions batch.
-        probabilities = model.predict(inputs_batch)[0]
+        # Get predictions from the model.
+        inputs = data.get_input_patch(year, (lon, lat), patch_size)
+        inputs_batch = np.stack([inputs])
+        probabilities = MODEL.predict(inputs_batch)[0]
+        predictions = probabilities.argmax(axis=-1).astype(np.uint8)
 
-        # Get discrete classification values from the probability distribution.
-        outputs = np.argmax(probabilities, axis=-1)
-
-        # Serialize the results into a compressed NumPy file.
-        with io.BytesIO() as f:
-            np.savez_compressed(f, inputs=patch, outputs=outputs)
-            return f.getvalue()
+        # Return the model predictions.
+        return {"predictions": predictions.tolist()}
 
     except Exception as e:
```

Comment:
ditto

Reply:
There could be many errors depending on the inputs for the request, or network issues. Anything can go wrong in Python and the process would just crash with the the error message staying in the server, and users would just get a generic 500 error. I wanted to make this into an isolation point so if anything goes wrong, we would still get a valid response with a readable error message on the client side. I'll add a comment.

I think this still falls into the [style guide](https://google.github.io/styleguide/pyguide.html#24-exceptions), but let me know if you prefer me to list all the exceptions I can find (it'll be a long list).

> * Never use catch-all except: statements, or catch Exception or StandardError, unless you are
>
>   * re-raising the exception, or
>   * _**creating an isolation point in the program where exceptions are not propagated but are recorded and suppressed instead, such as protecting a thread from crashing by guarding its outermost block.**_

Reply:
SGTM. Thanks for adding in the comment above!